aptos-move/framework/aptos-framework/sources/coin.spec.move

```diff
@@ -7,13 +11,245 @@ spec aptos_framework::coin {
         ensures [abstract] result.value == amount;
     }
 
-    spec coin_address {
+    spec coin_address<CoinType>(): address {
         pragma opaque;
+        aborts_if [abstract] false;
         ensures [abstract] result == spec_coin_address<CoinType>();
     }
 
     spec fun spec_coin_address<CoinType>(): address {
         // TODO: abstracted due to the lack of support for `type_info` in Prover.
```

Comment:
Now that Prover supports the `type_of` (https://github.com/move-language/move/pull/462), this can be specified.

Reply:
The specs should just reflect what the implementation does, so the question is about the implementation, seems not to abort on other system addresses.

Reply:
This PR updates the spec of `type_info` and has an example of how to use it in a spec: https://github.com/aptos-labs/aptos-core/pull/5341/files#diff-2f8d5cc984e8ad9aeddd19f91aeb2bb0a8421a8429399426631774b0218423b2R74

Reply:
I refer to https://github.com/move-language/move/blob/main/language/move-prover/tests/sources/functional/type_reflection.move

Reply:
First of all, could you rebase on the head of main if you haven't recently? There has been some updates.
And, could you explain more about Prover aborting at `type_of`?

Reply:
```
root@1eba4e23b8e5:/home/aptos/aptos-core# aptos -V
aptos 1.0.0
root@1eba4e23b8e5:/home/aptos/aptos-core# aptos move prove -V
aptos-move-prove 1.0.0
root@1eba4e23b8e5:/home/aptos/aptos-core# boogie /version 
Boogie program verifier version 2.15.8.0, Copyright (c) 2003-2014, Microsoft.
```

```
    public(friend) fun register_coin<CoinType>(account_addr: address) acquires Account {
        let account = borrow_global_mut<Account>(account_addr);
        event::emit_event<CoinRegisterEvent>(
            &mut account.coin_register_events,
            CoinRegisterEvent {
                type_info: type_info::type_of<CoinType>(),   // prover would abort at here 
            },
        );
    }
```

```
    ┌─ ./aptos-framework/sources/account.move:496:28
    │
496 │                 type_info: type_info::type_of<CoinType>(),
    │                            ------------------------------ abort happened here with execution failure
    │
    =     at ./aptos-framework/sources/coin.move:416: register
    =     at ./aptos-framework/sources/coin.spec.move:217
    =     at ./aptos-framework/sources/coin.spec.move:218
    =     at ./aptos-framework/sources/coin.move:416: register
    =         account = signer{0x4823}
    =     at ./aptos-framework/sources/coin.move:417: register
    =     at ./aptos-framework/../aptos-stdlib/../move-stdlib/sources/signer.move:12: address_of
    =         s = signer{0x4823}
    =     at ./aptos-framework/../aptos-stdlib/../move-stdlib/sources/signer.move:13: address_of
    =         result = 0x4823
    =     at ./aptos-framework/../aptos-stdlib/../move-stdlib/sources/signer.move:14: address_of
    =         account_addr = 0x4823
    =     at ./aptos-framework/sources/coin.move:419: register
    =     at ./aptos-framework/sources/coin.move:167: is_account_registered
    =         account_addr = 0x4823
    =     at ./aptos-framework/sources/coin.move:168: is_account_registered
    =         result = false
    =     at ./aptos-framework/sources/coin.move:169: initialize_with_parallelizable_supply
    =     at ./aptos-framework/sources/coin.move:418: register
    =     at ./aptos-framework/sources/coin.move:423: register
    =     at ./aptos-framework/sources/account.move:491: register_coin
    =         account_addr = 0x4823
    =     at ./aptos-framework/sources/account.move:492: register_coin
    =         account =
    =           &account.Account{
    =             authentication_key = vector{(size): 15724, default: 2u8},
    =             sequence_number = 6334,
    =             guid_creation_num = 18446744073709551613,
    =             coin_register_events =
    =               event.EventHandle{
    =                 counter = 19169,
    =                 guid =
    =                   guid.GUID{
    =                     id = guid.ID{creation_num = 24464, addr = 0x72ae}}},
    =             key_rotation_events =
    =               event.EventHandle{
    =                 counter = 26500,
    =                 guid =
    =                   guid.GUID{
    =                     id = guid.ID{creation_num = 26962, addr = 0x2cd6}}},
    =             rotation_capability_offer =
    =               account.CapabilityOffer{for = option.Option{vec = vector{}}},
    =             signer_capability_offer =
    =               account.CapabilityOffer{for = option.Option{vec = vector{}}}}
    =     at ./aptos-framework/sources/account.move:494: register_coin
    =     at ./aptos-framework/sources/account.move:496: register_coin
    =         ABORTED
```

Reply:
The head of main is up to date.

Reply:
OK. Thank you for the log. I will investigate it separately.